project/src/main/java/com/google/sps/data/Places.java

```diff
@@ -38,5 +44,38 @@
     return ImmutableList.copyOf(mutablePlaces);
   }
 
+  /**
+   * filters the given list of places according to the given parameters.
+   * @param places the list we want to sort.
+   * @param minRating The minimal rating we want to limit the list by. A number between 1 and 5.
+   * @param filterIfNoWebsite Specifies whether we should filter if there is no available website
+   *        for the restaurant.
+   * @param filterBranchesOfSamePlace Specifies whether we should remove different branches of the
+   *        same place from the results. Places are considered the same if they have the same name.
+   * @return the filtered list.
+   */
+  public static ImmutableList<Place> filter(ImmutableList<Place> places, int minRating, Boolean
+      filterIfNoWebsite, Boolean filterBranchesOfSamePlace) {
+    //TODO(M1): also filter by place's status when the attribute is added
+    ImmutableList<Place> result = ImmutableList.copyOf(
+        places.stream()
+            .filter(place -> place.rating() >= minRating)
+            .filter(filterIfNoWebsite
+                ? place -> !Strings.isNullOrEmpty(place.websiteUrl())
+                : place -> true)
+            .filter(filterBranchesOfSamePlace
+                ? distinctByKey(p -> p.name())
+                : p -> true)
+            .collect(Collectors.toList())
+    );
+    return result;
+  }
+
+  //A utility function that allows to distinct beteen objects using a certain property
+  private static <T> Predicate<T> distinctByKey(Function<? super T, Object> keyExtractor) {
+      Map<Object, Boolean> map = new ConcurrentHashMap<>();
+      return t -> map.putIfAbsent(keyExtractor.apply(t), Boolean.TRUE) == null;
```

Comment:
Why do we need to keep TRUE as value for every key in the map? Maybe it should just be a set and not a map if all values are always TRUE?

Reply:
Actually I didn't write the function, I used it after I saw a lot of places that recommended using it. It's possible that your improvement is good, but I think I'll keep it that way because I know it works well. What do you think?

Reply:
I just don't understand what this TRUE value is used for but I am fine with leaving it like this if ypu prefer :)

Reply:
Got rid of the function, so I'm resolving :)